filebeat/input/filestream/prospector.go

```diff
@@ -41,47 +38,63 @@ const (
 // The FS events then trigger either new Harvester runs or updates
 // the statestore.
 type fileProspector struct {
-	filewatcher  loginp.FSWatcher
-	identifier   fileIdentifier
-	ignoreOlder  time.Duration
-	cleanRemoved bool
+	filewatcher       loginp.FSWatcher
+	identifier        fileIdentifier
+	ignoreOlder       time.Duration
+	cleanRemoved      bool
+	stateChangeCloser stateChangeCloserConfig
 }
 
-func newFileProspector(
-	paths []string,
-	ignoreOlder time.Duration,
-	fileWatcherNs, identifierNs *common.ConfigNamespace,
-) (loginp.Prospector, error) {
+func (p *fileProspector) Init(cleaner loginp.ProspectorCleaner) error {
+	if p.cleanRemoved {
+		cleaner.CleanIf(func(v loginp.Value) bool {
+			var fm fileMeta
+			err := v.UnpackCursorMeta(&fm)
+			if err != nil {
+				// remove faulty entries
+				return true
+			}
 
-	filewatcher, err := newFileWatcher(paths, fileWatcherNs)
-	if err != nil {
-		return nil, err
+			_, err = os.Stat(fm.Source)
+			if err != nil {
+				return true
+			}
+			return false
+		})
 	}
 
-	identifier, err := newFileIdentifier(identifierNs)
-	if err != nil {
-		return nil, err
-	}
+	files := p.filewatcher.GetFiles()
```

Comment:
nice.

How about us using the list of known files in `CleanIf` as well. This way `CleanIf` will only remove entries in the registry that actually match the glob pattern. Having this I would feel more comfortable to remove the `!s.identifier.configuredUserID` check in the `CleanIf` implementation, such that `clean_remove` can be applied between restarts as well.

Reply:
I have removed the check you have mentioned from `CleanIf`. However, I have kept `MatchesInput` as a pre-filter to decrease the number of unnecessary lookups in the `files` map.